Hops 1157

smkniazi opened the pull request:
## Make sure there is no duplicate PR for this issue

* **Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added and passed (for bug fixes / features)
- [x] HOPS JIRA issue has been opened for this PR
- [x] All commits have been squashed down to a single commit
- [x] The commit message has the following format: [HOPS-XXX] message

* **Post a link to the associated JIRA issue**
https://hopshadoop.atlassian.net/projects/HOPS/issues/HOPS-1157

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)


* **What is the new behavior (if this is a feature change)?**


* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:
https://github.com/hopshadoop/hops-metadata-dal-impl-ndb/pull/197

HopsworksJenkins commented:
Can one of the admins verify this patch?

berthoug commented:
Why did you check the box: All commits have been squashed down to a single commit?

smkniazi commented:
test it

berthoug commented:
tests: https://bbc1.sics.se/jenkins/view/pull-manual/job/hops-pull-manual/73/

smkniazi commented:
TestGetBlocks is fixed now. Other failing tests are flaky and pass on my machine.

berthoug commented:
tests: https://bbc1.sics.se/jenkins/view/pull-manual/job/hops-pull-manual/74/

berthoug commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java:
```diff
@@ -2623,25 +2694,25 @@ private ReportStatistics reportDiff(final DatanodeStorageInfo storage,
-            .getAllStorageReplicasInBuckets(matchingResult.mismatchedBuckets);
+
+    final Map<Long, Long> mismatchedBlocksAndInodes = replicasInBucketsMT(storage, mismatchedBuckets);
+//    final Map<Long, Long> mismatchedBlocksAndInodes =  storage.getAllStorageReplicasInBuckets(mismatchedBuckets);
```
remove commented out code

berthoug commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPOfferService.java:
```diff
@@ -864,81 +846,41 @@ private void whirlingLikeASufi()
+      blkReportHander.processCommand(new DatanodeCommand[]{cmd});
     }
+
+//    if (brFailed){
```
remove commented out code

berthoug commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPOfferService.java:
```diff
@@ -1163,14 +1105,60 @@ void notifyNamenodeDeletedBlockInt(
       // block reporting has finished for now.
       if (blkReportHander != null) {
-        getLeaderActor().blockReportCompleted(bpRegistration);
+        BPServiceActor leader =  getLeaderActor();
```
shouldn't we retry if lead is null?

smkniazi replied:
The check is already there but it does not show here on github

berthoug commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPOfferService.java:
```diff
@@ -1163,14 +1105,60 @@ void notifyNamenodeDeletedBlockInt(
+
+  public static void removeMatchingBuckets(List<Integer> mismatchingBuckets, BlockReport report){
+    for(int i = 0; i < report.getBuckets().length; i++){
+      if(!mismatchingBuckets.contains(i)){
```
contains on a list is not efficient, the type of mismatchingBuckets should probably be a set.

smkniazi replied:
mismatchingBuckets is expected to contain only few elements. Only at cluster restart it will have at most 1000 elements. Set will not help, but HashSet will.

berthoug commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPOfferService.java:
```diff
@@ -1163,14 +1105,60 @@ void notifyNamenodeDeletedBlockInt(
+    Map<String, List<Integer>> map = ((HashesMismatchCommand) cmd).getMissMatchingBuckets();
+    for(StorageBlockReport report : reports){
+      List<Integer> mismatchingBuckets = map.get(report.getStorage().getStorageID());
+      if(mismatchingBuckets != null && mismatchingBuckets.size()>0){
```
won't this make that if all the bucket are matching it sends a full block report?

smkniazi replied:
no

berthoug replied:
You will need to explain more than that.
The way I read it the code does the following:
get a report for a storage, that I assume contains all the bucket.
get the mismatching bucket for the storage associated to this report.
if the number of mismatching buckets is more than 0:
   go through all the buckets in the report, mark to skip and set an empty block list for all the buckets that are not mismatching.

So if there are no mismatching buckets none of the matching buckets will be marked as skip and it will send all the blocks (full block report).
What am I missing?

berthoug commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java:
```diff
@@ -681,17 +681,23 @@ public boolean isSecurityEnabled() {
+      try {
+        leaderElection.stopElectionThread();
+      } catch (InterruptedException e) {
+        e.printStackTrace();
```
use the logger If you want to print something.

smkniazi replied:
done

berthoug commented on hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/protocol/DatanodeProtocol.java:
```diff
@@ -81,6 +81,7 @@
   final static int DNA_BALANCERBANDWIDTHUPDATE = 8; // update balancer bandwidth
   final static int DNA_CACHE = 9;      // cache blocks
   final static int DNA_UNCACHE = 10;   // uncache blocks
+  final static int DNA_HASHMISMATCH = 11;   // uncache blocks
```
change the comment to correspond to what it is.

smkniazi replied:
What?

berthoug replied:
is DNA_HASHMISMATCH = uncache blocks ?

berthoug commented on hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/TestBlockReport2.java:
```diff
@@ -1086,32 +1092,36 @@ private void matchDNandNNState(int nnId, int numDataNodes, MiniDFSCluster
         for (HashBucket storageHash : storageHashes) {
-          nnHashes.add(storageHash.getBucketId(), storageHash.getHash());
+          LOG.warn("XXX adding "+storageHash.getBucketId()+", "+ storageHash.getHash());
+//          nnHashes.add(storageHash.getBucketId(), storageHash.getHash());
```
remove commented out code

berthoug commented on hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/TestBlockReport2.java:
```diff
@@ -89,13 +91,13 @@ private void waitForTempReplica(MiniDFSCluster cluster, Block bl, int DN_N1)
     final int TIMEOUT = 40000;
 
     if (LOG.isDebugEnabled()) {
-      LOG.debug("Wait for datanode " + DN_N1 + " to appear");
```
is changing all the log.debug into log.info really needed?

smkniazi replied:
It is a unit test. It does not matter

berthoug replied:
Is this test only for Hops? If yes it does not matter, if no changing lines of codes for no real purpose like that is a source of conflict when merging hadoop jiras.

berthoug commented on hops-leader-election/src/main/java/io/hops/leaderElection/experiments/LightWeightNameNode.java:
```diff
@@ -56,7 +56,11 @@ public boolean isLeader() {
+      try {
+        leaderElection.stopElectionThread();
+      } catch (InterruptedException e) {
+        e.printStackTrace();
```
Could we have a log here instead of directly printing the stackTrace?

berthoug commented on hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/TestBlockReport2.java:
```diff
@@ -1435,4 +1449,82 @@ public void blockReport_15() throws IOException, InterruptedException {
+          blocksWritten++;
+        }
+      }catch (IOException e){
+       e.printStackTrace();
```
Could you use the logger?

smkniazi replied:
fixed

berthoug commented:
I don't manage to run the following tests neither on the test machine nore on my own machine, can you have a look at them:
TestBalancer
TestDecommission

berthoug commented:
Forget about TestDecommission I also don't manage to make it run on master. But testBalancer run at first try while I did not manage to have it run on your code after at least 10 tries.

smkniazi commented:
In `TestDecommission`,  testNodesPerInterval and testBlocksPerInterval are flaky. I have added 500 ms sleep and the tests passed five times in a row. 

TestBalancer was broken before this jira. The issue was that one of the unit test did not close the MiniDfscluster object. I have fixed it was ran the tests. It passes.

berthoug commented:
The two tests are still not passing at all on the test machine. I will test on my machine and come back to you.

berthoug commented:
Can you resolve the conflicts?

berthoug commented:
Are you sure of what you pushed? It does not seem to build: https://bbc1.sics.se/jenkins/view/pull-manual/job/hops-pull-manual/86/console

smkniazi commented:
should work now. Forgot to push one class

berthoug commented:
tests: https://bbc1.sics.se/jenkins/view/pull-manual/job/hops-pull-manual/89/
